File: packages/python/elizaos/features/basic_capabilities/providers/knowledge.py
from __future__ import annotations

from typing import TYPE_CHECKING

from elizaos.generated.spec_helpers import require_provider_spec
from elizaos.types import Provider, ProviderResult

if TYPE_CHECKING:
    from elizaos.types import IAgentRuntime, Memory, State

# Get text content from centralized specs
_spec = require_provider_spec("KNOWLEDGE")


async def get_knowledge_context(
    runtime: IAgentRuntime,
    message: Memory,
    state: State | None = None,
) -> ProviderResult:
    sections: list[str] = []
    knowledge_entries: list[dict[str, str]] = []

    query_text = ""
    if message.content and message.content.text:
        query_text = message.content.text

    if not query_text:
        return ProviderResult(
            text="", values={"knowledgeCount": 0, "hasKnowledge": False}, data={"entries": []}
        )

    # 1. Fetch recent messages to get embeddings
    recent_messages = await runtime.get_memories(
        room_id=message.room_id, limit=5, table_name="messages"
    )

    # 2. Extract valid embeddings
    embeddings = [m.embedding for m in recent_messages if m and m.embedding]

    relevant_knowledge = []
    # 3. Search using the most recent embedding if available
    if embeddings:
        primary_embedding = embeddings[0]
        params = {
            "tableName": "knowledge",
            "roomId": str(message.room_id),
            "embedding": primary_embedding,
            "matchThreshold": 0.75,
            "matchCount": 5,
            "unique": True,
        }
        relevant_knowledge = await runtime.search_memories(params)
    elif query_text:
        # Fallback to search_knowledge if no embeddings found?
        # TS implementation might rely on search_memories logic handling missing embedding?
        # No, TS skips if no embedding.
        pass

    for entry in relevant_knowledge:
        if entry.content and entry.content.text:
            knowledge_text = entry.content.text
            if len(knowledge_text) > 500:
                knowledge_text = knowledge_text[:500] + "..."

            entry_dict = {
                "id": str(entry.id) if entry.id else "",
                "text": knowledge_text,
                "source": str(entry.metadata.get("source", "unknown"))
                if entry.metadata
                else "unknown",
            }
            knowledge_entries.append(entry_dict)
            sections.append(f"- {knowledge_text}")

    context_text = "# Relevant Knowledge\n" + "\n".join(sections) if sections else ""

    return ProviderResult(
        text=context_text,
        values={
            "knowledgeCount": len(knowledge_entries),
            "hasKnowledge": len(knowledge_entries) > 0,
        },
        data={
            "entries": knowledge_entries,
            "query": query_text,
        },
    )


knowledge_provider = Provider(
    name=_spec["name"],
    description=_spec["description"],
    get=get_knowledge_context,
    dynamic=_spec.get("dynamic", True),
    position=_spec.get("position"),
)

File: packages/rust/src/features/advanced_capabilities/providers/knowledge.rs
//! KNOWLEDGE provider implementation.

use async_trait::async_trait;
use once_cell::sync::Lazy;

use crate::error::PluginResult;
use crate::generated::spec_helpers::require_provider_spec;
use crate::runtime::IAgentRuntime;
use crate::types::database::SearchMemoriesParams;
use crate::types::{Memory, MemoryType, ProviderResult, State};

use super::Provider;

static SPEC: Lazy<&'static crate::generated::spec_helpers::ProviderDoc> =
    Lazy::new(|| require_provider_spec("KNOWLEDGE"));

/// Provider for knowledge information.
pub struct KnowledgeProvider;

#[async_trait]
impl Provider for KnowledgeProvider {
    fn name(&self) -> &'static str {
        &SPEC.name
    }

    fn description(&self) -> &'static str {
        &SPEC.description
    }

    fn is_dynamic(&self) -> bool {
        SPEC.dynamic.unwrap_or(true)
    }

    async fn get(
        &self,
        runtime: &dyn IAgentRuntime,
        message: &Memory,
        _state: Option<&State>,
    ) -> PluginResult<ProviderResult> {
        let query_text = &message.content.text;

        if query_text.is_empty() {
            return Ok(ProviderResult::new("")
                .with_value("knowledgeCount", 0)
                .with_value("hasKnowledge", false)
                .with_data("entries", Vec::<serde_json::Value>::new()));
        }

        let mut sections = Vec::new();
        let mut knowledge_entries: Vec<serde_json::Value> = Vec::new();

        // 1. Fetch recent messages to get embeddings
        let recent_messages = match runtime
            .get_memories(Some(message.room_id), None, None, 5)
            .await
        {
            Ok(msgs) => msgs,
            Err(e) => {
                runtime.log_warning(
                    "provider:knowledge",
                    &format!("Error fetching recent messages: {}", e),
                );
                Vec::new()
            }
        };

        // 2. Extract valid embeddings
        let embeddings: Vec<Vec<f32>> = recent_messages
            .into_iter()
            .filter_map(|m| m.embedding)
            .filter(|v| !v.is_empty())
            .collect();

        // 3. Search using the most recent embedding if available
        if let Some(primary_embedding) = embeddings.first() {
            let params = SearchMemoriesParams {
                table_name: "knowledge".to_string(),
                room_id: message.room_id,
                embedding: primary_embedding.clone(),
                match_threshold: 0.75,
                match_count: 5,
                unique: Some(true),
            };

            match runtime.search_memories(params).await {
                Ok(entries) => {
                    for entry in entries {
                        if !entry.content.text.is_empty() {
                            let mut knowledge_text = entry.content.text.clone();
                            if knowledge_text.len() > 500 {
                                knowledge_text.truncate(500);
                                knowledge_text.push_str("...");
                            }

                            let source = entry
                                .metadata
                                .get("source")
                                .and_then(|v| v.as_str())
                                .unwrap_or("unknown")
                                .to_string();

                            // Avoid duplicates
                            let id_str = entry.id.to_string();
                            if !knowledge_entries.iter().any(|e| e["id"] == id_str) {
                                knowledge_entries.push(serde_json::json!({
                                    "id": id_str,
                                    "text": knowledge_text,
                                    "source": source
                                }));
                                sections.push(format!("- {}", knowledge_text));
                            }
                        }
                    }
                }
                Err(e) => {
                     runtime.log_warning(
                        "provider:knowledge",
                        &format!("Error searching knowledge: {}", e),
                    );
                }
            }
        }

        let context_text = if sections.is_empty() {
            String::new()
        } else {
            format!("# Relevant Knowledge\n{}", sections.join("\n"))
        };

        Ok(ProviderResult::new(context_text)
            .with_value("knowledgeCount", knowledge_entries.len() as i64)
            .with_value("hasKnowledge", !knowledge_entries.is_empty())
            .with_data("entries", knowledge_entries)
            .with_data("query", query_text.clone()))
    }
}
